# Práctica 1 — Análisis Exploratorio de Datos (EDA)
## Tema: Ciberseguridad

Este notebook realiza un **Análisis Exploratorio de Datos (EDA)** sobre tres fuentes de datos relacionadas con ciberseguridad:

1. **`cybersecurity_threats`** — tabla en base de datos PostgreSQL con amenazas e incidentes de seguridad.
2. **`Cyber_security.csv`** — archivo CSV con índices globales de ciberseguridad por país (GCI, NCSI, CEI, DDL).
3. **`internet_users_by_country_cleaned.csv`** — archivo CSV con estadísticas de usuarios de internet por país.

---
### Objetivo del EDA
El Análisis Exploratorio de Datos permite:
- Conocer la **estructura y dimensiones** de los datos.
- Identificar los **tipos de datos** de cada columna.
- Detectar **valores nulos o faltantes**.
- Obtener **estadísticas descriptivas** (media, máximo, mínimo, etc.).
- Realizar **agrupaciones** para identificar patrones por categorías.

---
## SECCIÓN 1: Importación de Librerías y Configuración del Entorno

### Paso 1 — Importar librerías y cargar variables de entorno

**¿Qué hace esta celda?**

Se importan las librerías necesarias para el proyecto:
- **`pandas`**: librería principal para manipulación y análisis de datos en Python. Permite trabajar con estructuras de datos tipo tabla llamadas `DataFrame`.
- **`dotenv` / `load_dotenv()`**: permite cargar variables de entorno desde un archivo `.env`. Esto es una buena práctica de seguridad: las credenciales de la base de datos (usuario, contraseña, nombre de la BD) no se escriben directamente en el código, sino que se guardan en un archivo `.env` que no se sube al repositorio.
- **`os`**: módulo de Python para interactuar con el sistema operativo, en este caso para leer las variables de entorno cargadas.
- **`sqlalchemy` / `create_engine`**: librería que actúa como puente entre Python y bases de datos relacionales. `create_engine` crea la conexión a la base de datos.

In [1]:
import pandas as pd
from dotenv import load_dotenv
import os
load_dotenv()
from sqlalchemy import create_engine

### Paso 2 — Leer las credenciales de la base de datos desde el entorno

**¿Qué hace esta celda?**

Se leen las credenciales de conexión a la base de datos PostgreSQL desde las variables de entorno definidas en el archivo `.env`. Las variables son:
- **`POSTGRES_USER`**: nombre del usuario de la base de datos.
- **`POSTGRES_PASSWORD`**: contraseña del usuario.
- **`POSTGRES_DB`**: nombre de la base de datos a la que se conectará.
- **`POSTGRES_HOST`**: dirección del servidor donde corre la base de datos (en este caso `localhost`, es decir, la máquina local).

> **Buena práctica**: Nunca se deben escribir contraseñas directamente en el código. El uso de `.env` protege las credenciales y facilita cambiarlas sin modificar el código fuente.

In [2]:
DB_USER = os.getenv("POSTGRES_USER")
DB_PASSWORD = os.getenv("POSTGRES_PASSWORD")
DB_NAME = os.getenv("POSTGRES_DB")
DB_HOST = os.getenv("POSTGRES_HOST")

### Paso 3 — Crear el motor de conexión a PostgreSQL

**¿Qué hace esta celda?**

Se crea el **motor de conexión** (`engine`) a la base de datos PostgreSQL usando SQLAlchemy. La cadena de conexión tiene el formato:

```
postgresql+psycopg2://usuario:contraseña@host/nombre_bd
```

- **`postgresql+psycopg2`**: indica que se usará PostgreSQL como motor de base de datos y `psycopg2` como driver de Python para conectarse.
- Las variables `DB_USER`, `DB_PASSWORD`, `DB_NAME` son las credenciales leídas en el paso anterior.
- **`localhost`**: indica que la base de datos corre en la misma máquina donde se ejecuta el notebook.

Este `engine` se utilizará en el siguiente paso para ejecutar consultas SQL y cargar los datos en un DataFrame de pandas.

In [3]:
engine = create_engine(f'postgresql+psycopg2://{DB_USER}:{DB_PASSWORD}@localhost/{DB_NAME}')


---
## SECCIÓN 2: Carga de Datos

### Paso 4 — Cargar datos desde la base de datos PostgreSQL

**¿Qué hace esta celda?**

Se ejecuta una consulta SQL (`SELECT * FROM cybersecurity_threats`) directamente desde Python usando `pd.read_sql()`. Esta función envía la consulta a la base de datos a través del `engine` creado anteriormente y carga el resultado directamente en un **DataFrame** de pandas llamado `df_threads`.

- **`SELECT * FROM cybersecurity_threats`**: selecciona todos los registros y columnas de la tabla `cybersecurity_threats`.
- El resultado queda disponible como un DataFrame listo para análisis.

> **Nota sobre el error**: En la ejecución registrada en este notebook, la celda produjo un `OperationalError` porque el servidor PostgreSQL no estaba activo en el momento de la ejecución (`Connection refused` en el puerto 5432). Esto es un error de entorno (la base de datos no estaba corriendo), no un error en el código. Para ejecutarla correctamente, se debe tener el servidor PostgreSQL iniciado y accesible en `localhost:5432`.

In [5]:
df_threads=pd.read_sql('select * from cybersecurity_threats', engine)

### Paso 5 — Cargar el dataset de índices de ciberseguridad (CSV)

**¿Qué hace esta celda?**

Se carga el archivo `Cyber_security.csv` desde la carpeta `data/` usando `pd.read_csv()`. Este archivo contiene índices globales de ciberseguridad por país. El DataFrame resultante se llama `df_cib_EDA`.

Las columnas del dataset son:
- **`Country`**: nombre del país.
- **`Region`**: región geográfica (Africa, Europe, Asia-Pasific, etc.).
- **`CEI`** *(Cybersecurity Exposure Index)*: índice de exposición a amenazas cibernéticas (entre 0 y 1; valores más altos = mayor exposición).
- **`GCI`** *(Global Cybersecurity Index)*: índice global de ciberseguridad de la ITU (entre 0 y 100; valores más altos = mayor madurez en ciberseguridad).
- **`NCSI`** *(National Cyber Security Index)*: índice nacional de capacidad en ciberseguridad.
- **`DDL`** *(Digital Development Level)*: nivel de desarrollo digital del país.

In [6]:
df_cib_EDA = pd.read_csv('data/Cyber_security.csv')

### Paso 6 — Cargar el dataset de usuarios de internet por país (CSV)

**¿Qué hace esta celda?**

Se carga el archivo `internet_users_by_country_cleaned.csv` desde la carpeta `data/`. Este es un dataset ya **limpiado** ("cleaned"), lo que indica que pasó por un proceso de preprocesamiento previo. El DataFrame resultante se llama `df_cib_usenet`.

Las columnas del dataset son:
- **`Country or area`**: nombre del país o territorio.
- **`Subregion`**: subregión geográfica (Eastern Asia, Southern Asia, etc.).
- **`Region`**: región geográfica principal (Asia, Americas, Europe, etc.).
- **`Internet users`**: número absoluto de usuarios de internet.
- **`Population_2021`**: población total del país en el año 2021.
- **`Year`**: año al que corresponden los datos de usuarios de internet.
- **`Percentage_Internet_Users`**: porcentaje de la población que usa internet.

In [7]:
df_cib_usenet = pd.read_csv('data/internet_users_by_country_cleaned.csv')

---
## SECCIÓN 3: EDA — Dataset `df_threads` (Amenazas de Ciberseguridad - PostgreSQL)

> **Nota**: Las celdas de esta sección requieren que `df_threads` esté correctamente cargado desde la base de datos (Paso 4). Dado que la conexión a PostgreSQL falló en la ejecución registrada, estas celdas muestran `NameError`. El código es correcto y funcionará cuando la base de datos esté disponible.

### Paso 7 — Visualizar las primeras filas (`head()`)

**¿Qué hace esta celda?**

El método `head()` muestra las **primeras 5 filas** del DataFrame (por defecto). Es el primer paso de exploración para tener una vista rápida de cómo están estructurados los datos: qué columnas existen, qué tipo de valores contienen y cómo se ven los registros reales.

Permite responder: *¿Qué forma tienen mis datos? ¿Tiene sentido lo que veo?*

In [8]:
df_threads.head()

,id,country,year,attack_type,target_industry,financial_loss_in_million_usd,number_of_affected_users,attack_source,security_vulnerability_type,defense_mechanism_used,incident_resolution_time_in_hours
0,1,China,2019,Phishing,Education,80.53,773169,Hacker Group,Unpatched Software,VPN,63
1,2,China,2019,Ransomware,Retail,62.19,295961,Hacker Group,Unpatched Software,Firewall,71
2,3,India,2017,Man-in-the-Middle,IT,38.65,605895,Hacker Group,Weak Passwords,VPN,20
3,4,UK,2024,Ransomware,Telecommunications,41.44,659320,Nation-state,Social Engineering,AI-based Detection,7
4,5,Germany,2018,Man-in-the-Middle,IT,74.41,810682,Insider,Social Engineering,VPN,68


### Paso 8 — Conocer las dimensiones del DataFrame (`shape`)

**¿Qué hace esta celda?**

El atributo `shape` devuelve una tupla `(filas, columnas)` que indica la **dimensión exacta** del DataFrame. Por ejemplo, `(1000, 15)` significaría 1000 registros y 15 columnas.

Es fundamental conocer el tamaño del dataset antes de cualquier análisis para saber con cuánta información se está trabajando.

In [9]:
df_threads.shape

(3000, 11)

### Paso 9 — Revisar los tipos de datos de cada columna (`dtypes`)

**¿Qué hace esta celda?**

El atributo `dtypes` muestra el **tipo de dato** asignado a cada columna del DataFrame. Los tipos más comunes son:
- **`object`** / **`str`**: texto o cadenas de caracteres.
- **`int64`**: números enteros.
- **`float64`**: números decimales.
- **`datetime64`**: fechas y tiempos.

Verificar los tipos de datos es clave porque determina qué operaciones se pueden aplicar a cada columna. Por ejemplo, no se puede calcular la media de una columna de texto, y las fechas cargadas como texto requieren conversión antes de hacer cálculos temporales.

In [10]:
df_threads.dtypes

id                                     int64
country                                  str
year                                   int64
attack_type                              str
target_industry                          str
financial_loss_in_million_usd        float64
number_of_affected_users               int64
attack_source                            str
security_vulnerability_type              str
defense_mechanism_used                   str
incident_resolution_time_in_hours      int64
dtype: object

### Paso 10 — Resumen estructural completo del DataFrame (`info()`)

**¿Qué hace esta celda?**

`info()` proporciona un **resumen completo** del DataFrame en un solo vistazo:
- Número total de filas (RangeIndex).
- Lista de columnas con su tipo de dato.
- Cantidad de **valores no nulos** por columna (crucial para detectar datos faltantes).
- Uso de memoria del DataFrame.

Es una de las funciones más útiles en la fase inicial del EDA porque combina la información de `shape` y `dtypes`, y además añade la cuenta de valores no nulos para detectar columnas con datos faltantes de forma rápida.

In [11]:
df_threads.info()

<class 'pandas.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 11 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   id                                 3000 non-null   int64  
 1   country                            3000 non-null   str    
 2   year                               3000 non-null   int64  
 3   attack_type                        3000 non-null   str    
 4   target_industry                    3000 non-null   str    
 5   financial_loss_in_million_usd      3000 non-null   float64
 6   number_of_affected_users           3000 non-null   int64  
 7   attack_source                      3000 non-null   str    
 8   security_vulnerability_type        3000 non-null   str    
 9   defense_mechanism_used             3000 non-null   str    
 10  incident_resolution_time_in_hours  3000 non-null   int64  
dtypes: float64(1), int64(4), str(6)
memory usage: 257.9 KB


### Paso 11 — Estadísticas descriptivas de las columnas numéricas (`describe()`)

**¿Qué hace esta celda?**

`describe()` genera un resumen estadístico de todas las **columnas numéricas** del DataFrame. Las métricas que calcula son:
- **`count`**: cantidad de valores no nulos.
- **`mean`**: promedio.
- **`std`**: desviación estándar (dispersión de los datos).
- **`min`** / **`max`**: valor mínimo y máximo.
- **`25%`, `50%`, `75%`**: percentiles (el 50% es la mediana).

Permite detectar rápidamente valores atípicos (outliers), la distribución general de los datos y si los rangos de los valores son razonables para el contexto.

In [12]:
df_threads.describe()

,id,year,financial_loss_in_million_usd,number_of_affected_users,incident_resolution_time_in_hours
count,3000.000000,3000.000000,3000.000000,3000.000000,3000.000000
mean,1500.500000,2019.570333,50.492970,504684.136333,36.476000
std,866.169729,2.857932,28.791415,289944.084972,20.570768
min,1.000000,2015.000000,0.500000,424.000000,1.000000
25%,750.750000,2017.000000,25.757500,255805.250000,19.000000
50%,1500.500000,2020.000000,50.795000,504513.000000,37.000000
75%,2250.250000,2022.000000,75.630000,758088.500000,55.000000
max,3000.000000,2024.000000,99.990000,999635.000000,72.000000


### Paso 12 — Detectar si existe algún valor nulo por columna (`isnull().any()`)

**¿Qué hace esta celda?**

`isnull()` genera un DataFrame booleano donde cada celda es `True` si el valor es nulo (`NaN`) y `False` si no lo es. Al encadenar `.any()`, se colapsa ese resultado por columna: si **algún** valor de la columna es nulo, el resultado es `True`; si ninguno es nulo, el resultado es `False`.

Esto da una respuesta rápida a la pregunta: *¿Alguna columna tiene datos faltantes?* Es el primer filtro de calidad de datos.

In [13]:
df_threads.isnull().any()

id                                   False
country                              False
year                                 False
attack_type                          False
target_industry                      False
financial_loss_in_million_usd        False
number_of_affected_users             False
attack_source                        False
security_vulnerability_type          False
defense_mechanism_used               False
incident_resolution_time_in_hours    False
dtype: bool

### Paso 13 — Contar exactamente cuántos valores nulos hay por columna (`isnull().sum()`)

**¿Qué hace esta celda?**

Mientras `.any()` solo indica si existe al menos un nulo, `.sum()` cuenta **exactamente cuántos valores nulos** hay en cada columna. Dado que `True` vale 1 y `False` vale 0, sumar los valores booleanos equivale a contar los nulos.

Esta información es esencial para decidir la estrategia de tratamiento de datos faltantes: si hay muy pocos nulos, se pueden eliminar esas filas; si son muchos, es mejor imputar valores (media, mediana, moda, etc.) o analizar por qué faltan esos datos.

In [14]:
df_threads.isnull().sum()

id                                   0
country                              0
year                                 0
attack_type                          0
target_industry                      0
financial_loss_in_million_usd        0
number_of_affected_users             0
attack_source                        0
security_vulnerability_type          0
defense_mechanism_used               0
incident_resolution_time_in_hours    0
dtype: int64

### Paso 14 — Estadísticas agregadas sobre la pérdida financiera (`agg()`)

**¿Qué hace esta celda?**

Se selecciona la columna `Financial Loss (in Million $)` y se calculan tres métricas de resumen con `.agg()`:
- **`mean`**: pérdida financiera promedio por incidente.
- **`max`**: máxima pérdida financiera registrada en un solo incidente.
- **`min`**: mínima pérdida financiera registrada.

Esto permite entender el **impacto económico** típico y extremo de los incidentes de ciberseguridad registrados en el dataset. `.agg()` es útil porque permite calcular múltiples funciones de agregación en una sola llamada.

In [17]:
df_threads['financial_loss_in_million_usd'].agg(['mean', 'max', 'min'])

mean    50.49297
max     99.99000
min      0.50000
Name: financial_loss_in_million_usd, dtype: float64

### Paso 15 — Análisis agrupado por industria objetivo (`groupby()`)

**¿Qué hace esta celda?**

Se realiza un **agrupamiento** por la columna `Target Industry` (industria objetivo del ataque) y se calculan el máximo y mínimo de dos variables para cada grupo:
- **`Financial Loss (in Million $)`**: pérdida financiera en millones de dólares.
- **`Incident Resolution Time (in Hours)`**: tiempo en horas que tardó en resolverse el incidente.

Esto permite comparar qué industrias sufren mayores pérdidas económicas y cuáles tardan más tiempo en resolver los incidentes de ciberseguridad. Es un análisis categórico que revela patrones de impacto según el sector atacado.

`groupby()` es una de las operaciones más poderosas del EDA: divide los datos en grupos, aplica una función y combina los resultados.

In [21]:
df_threads.groupby('target_industry')[['financial_loss_in_million_usd']].agg(['max', 'min'])

financial_loss_in_million_usd      
                                             max   min
target_industry                                       
Banking                                    99.99  1.01
Education                                  99.83  0.54
Government                                 99.72  0.61
Healthcare                                 99.97  0.75
IT                                         99.99  1.01
Retail                                     99.78  0.50
Telecommunications                         99.98  0.54

---
## SECCIÓN 4: EDA — Dataset `df_cib_EDA` (Índices Globales de Ciberseguridad)

Este dataset contiene **192 países** con 6 columnas: `Country`, `Region`, `CEI`, `GCI`, `NCSI` y `DDL`. Las columnas numéricas tienen valores faltantes significativos (especialmente CEI con 84 nulos).

### Paso 16 — Visualizar las primeras filas (`head()`)

In [22]:
df_cib_EDA.head()

,Country,Region,CEI,GCI,NCSI,DDL
0,Afghanistan,Asia-Pasific,1.000,5.20,11.69,19.50
1,Albania,Europe,0.566,64.32,62.34,48.74
2,Algeria,Africa,0.721,33.95,33.77,42.81
3,Andorra,Europe,NaN,26.38,NaN,NaN
4,Angola,Africa,NaN,12.99,9.09,22.69


> **Observación del resultado**: Se pueden ver los primeros 5 países. Nótese que Andorra (fila 3) tiene `NaN` (Not a Number) en las columnas CEI, NCSI y DDL, lo que indica **datos faltantes**. Esto es normal en datasets reales y se cuantificará en los pasos de análisis de nulos.

### Paso 17 — Conocer las dimensiones del DataFrame (`shape`)

**¿Qué hace esta celda?** Devuelve la tupla `(filas, columnas)` para saber el tamaño exacto del dataset.

In [23]:
df_cib_EDA.shape

(192, 6)

> **Resultado**: `(192, 6)` → El dataset tiene **192 filas** (países) y **6 columnas**.

### Paso 18 — Revisar los tipos de datos de cada columna (`dtypes`)

**¿Qué hace esta celda?** Muestra el tipo de dato de cada columna.

In [24]:
df_cib_EDA.dtypes

Country        str
Region         str
CEI        float64
GCI        float64
NCSI       float64
DDL        float64
dtype: object

> **Resultado**: `Country` y `Region` son texto (`str`). Los índices CEI, GCI, NCSI y DDL son decimales (`float64`), lo que es correcto dado que son puntuaciones con valores decimales.

### Paso 19 — Resumen estructural completo del DataFrame (`info()`)

**¿Qué hace esta celda?** Muestra el resumen completo incluyendo la cuenta de valores no nulos.

In [29]:
df_cib_EDA.info()

<class 'pandas.DataFrame'>
RangeIndex: 192 entries, 0 to 191
Data columns (total 6 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Country  192 non-null    str    
 1   Region   192 non-null    str    
 2   CEI      108 non-null    float64
 3   GCI      190 non-null    float64
 4   NCSI     167 non-null    float64
 5   DDL      152 non-null    float64
dtypes: float64(4), str(2)
memory usage: 9.1 KB


> **Observación clave**: Solo `Country` y `Region` tienen los 192 valores completos. Las columnas numéricas tienen datos faltantes:
> - CEI: solo 108 de 192 países tienen dato (44% faltante).
> - GCI: 190 de 192 (casi completo).
> - NCSI: 167 de 192.
> - DDL: 152 de 192.

### Paso 20 — Estadísticas descriptivas de las columnas numéricas (`describe()`)

**¿Qué hace esta celda?** Genera el resumen estadístico de las columnas numéricas.

In [30]:
df_cib_EDA.describe()

              CEI         GCI        NCSI         DDL
count  108.000000  190.000000  167.000000  152.000000
mean     0.471861   52.769526   43.306587   51.707829
std      0.217246   34.884013   26.820907   18.283847
min      0.110000    1.350000    1.300000   11.300000
25%      0.279500   18.257500   19.480000   36.532500
50%      0.483000   53.145000   40.260000   51.790000
75%      0.624250   89.187500   64.940000   65.237500
max      1.000000  100.000000   96.100000   82.930000

> **Interpretación**:
> - **CEI** promedio de 0.47 (en escala 0-1): exposición moderada a amenazas en promedio global.
> - **GCI** promedio de 52.77 (en escala 0-100): nivel de madurez en ciberseguridad moderado a nivel mundial.
> - **NCSI** promedio de 43.31: capacidad nacional en ciberseguridad relativamente baja en promedio global.
> - La desviación estándar alta en GCI (34.88) indica gran disparidad entre países.

### Paso 21 — Detectar columnas con valores nulos (`isnull().any()`)

**¿Qué hace esta celda?** Indica con `True`/`False` si cada columna tiene al menos un valor nulo.

In [15]:
df_cib_EDA.isnull().any()

Country    False
Region     False
CEI         True
GCI         True
NCSI        True
DDL         True
dtype: bool

> **Resultado**: Las columnas `Country` y `Region` no tienen nulos. Las cuatro columnas de índices numéricos (CEI, GCI, NCSI, DDL) tienen al menos un valor faltante.

### Paso 22 — Contar exactamente los valores nulos por columna (`isnull().sum()`)

**¿Qué hace esta celda?** Cuenta el número exacto de nulos en cada columna.

In [16]:
df_cib_EDA.isnull().sum()

Country     0
Region      0
CEI        84
GCI         2
NCSI       25
DDL        40
dtype: int64

> **Resultado**: CEI tiene **84 nulos** (el 44% del dataset), lo que es significativo. GCI solo tiene 2 nulos. Esto puede influir en la representatividad del CEI para análisis comparativos regionales.

### Paso 23 — Estadísticas de la columna GCI (`agg()`)

**¿Qué hace esta celda?** Calcula la media, máximo y mínimo del Índice Global de Ciberseguridad (GCI).

In [17]:
df_cib_EDA['GCI'].agg(['mean', 'max', 'min'])

mean     52.769526
max     100.000000
min       1.350000
Name: GCI, dtype: float64

> **Interpretación**:
> - El **GCI promedio global** es de **52.77 sobre 100**, lo que indica que globalmente los países tienen un nivel de madurez en ciberseguridad apenas superior al 50%.
> - El máximo es 100 (país con máxima madurez, probablemente países como USA, UK o Estonia).
> - El mínimo es 1.35 (país con prácticamente nula infraestructura de ciberseguridad).
> - Esta brecha enorme (1.35 a 100) refleja la gran desigualdad global en ciberseguridad.

### Paso 24 — Análisis agrupado por región: GCI y NCSI (`groupby()`)

**¿Qué hace esta celda?** Agrupa los países por región geográfica y calcula el máximo y mínimo de los índices GCI y NCSI dentro de cada región.

In [18]:
df_cib_EDA.groupby('Region')[['GCI', 'NCSI']].agg(['max', 'min'])

                  GCI          NCSI       
                  max    min    max    min
Region                                    
Africa          96.89   1.46  70.13   1.30
Asia-Pasific    99.54   1.35  84.42   2.60
Europe          99.54  13.83  96.10   2.60
North America  100.00   2.20  71.43   3.90
South America   96.60  16.14  63.64  10.39

> **Interpretación**:
> - **North America** tiene el GCI máximo de 100 (el más alto globalmente).
> - **Europe** tiene el NCSI más alto (96.1), lo que refleja las políticas de ciberseguridad europeas (GDPR, ENS, etc.).
> - **Africa** tiene los valores mínimos más bajos tanto en GCI como NCSI, lo que indica menor desarrollo en ciberseguridad.
> - **South America** muestra los mínimos más altos en ambas métricas, sugiriendo mayor homogeneidad en la región (menos países con niveles críticos).

---
## SECCIÓN 5: EDA — Dataset `df_cib_usenet` (Usuarios de Internet por País)

Este dataset contiene **215 países** con 7 columnas. A diferencia del dataset anterior, **no tiene valores nulos** — es un dataset limpio y completo.

### Paso 25 — Visualizar las primeras filas (`head()`)

In [14]:
df_cib_usenet.head()

  Country or area           Subregion    Region  Internet users  \
0           China        Eastern Asia      Asia      1102140000   
1           India       Southern Asia      Asia       881250000   
2   United States    Northern America  Americas       311300000   
3       Indonesia  South-eastern Asia      Asia       215626156   
4        Pakistan       Southern Asia      Asia       170000000   

   Population_2021  Year  Percentage_Internet_Users  
0       1425893465  2022                       77.3  
1       1407563842  2024                       62.6  
2        336997624  2023                       92.4  
3        273753191  2023                       78.8  
4        240000000  2022                       70.8  

> **Observación**: El dataset está ordenado de mayor a menor número de usuarios de internet. China (1.1 mil millones) e India (881 millones) encabezan la lista. El dataset incluye tanto el número absoluto de usuarios como el porcentaje respecto a la población.

### Paso 26 — Conocer las dimensiones del DataFrame (`shape`)

In [32]:
df_cib_usenet.shape

(215, 7)

> **Resultado**: `(215, 7)` → El dataset tiene **215 filas** (países/territorios) y **7 columnas**. Tiene más registros que el dataset de ciberseguridad (192) porque incluye algunos territorios además de países reconocidos.

### Paso 27 — Revisar los tipos de datos de cada columna (`dtypes`)

In [33]:
df_cib_usenet.dtypes

Country or area                  str
Subregion                        str
Region                           str
Internet users                 int64
Population_2021                int64
Year                           int64
Percentage_Internet_Users    float64
dtype: object

> **Resultado**: Las columnas de texto son correctamente `str`. Los usuarios de internet y la población son `int64` (enteros), y el porcentaje es `float64` (decimal). Los tipos de datos son apropiados para cada variable.

### Paso 28 — Resumen estructural completo del DataFrame (`info()`)

In [34]:
df_cib_usenet.info()

<class 'pandas.DataFrame'>
RangeIndex: 215 entries, 0 to 214
Data columns (total 7 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Country or area            215 non-null    str    
 1   Subregion                  215 non-null    str    
 2   Region                     215 non-null    str    
 3   Internet users             215 non-null    int64  
 4   Population_2021            215 non-null    int64  
 5   Year                       215 non-null    int64  
 6   Percentage_Internet_Users  215 non-null    float64
dtypes: float64(1), int64(3), str(3)
memory usage: 11.9 KB


> **Resultado**: Todas las columnas tienen **215 valores no nulos** (completo). Este es un dataset limpio sin datos faltantes, lo que simplifica el análisis posterior.

### Paso 29 — Estadísticas descriptivas de las columnas numéricas (`describe()`)

In [35]:
df_cib_usenet.describe()

       Internet users  Population_2021         Year  Percentage_Internet_Users
count    2.150000e+02     2.150000e+02   215.000000                 215.000000
mean     2.464109e+07     3.668079e+07  2021.227907                  57.368372
std      1.015554e+08     1.418360e+08     0.647903                  29.269782
min      1.034000e+03     1.937000e+03  2020.000000                   1.700000
25%      3.506630e+05     8.130960e+05  2021.000000                  30.250000
50%      2.532059e+06     6.703799e+06  2021.000000                  64.100000
75%      1.001494e+07     2.558691e+07  2021.000000                  82.000000
max      1.102140e+09     1.425893e+09  2024.000000                 100.000000

> **Interpretación**:
> - El **porcentaje promedio** de usuarios de internet es **57.37%** a nivel mundial.
> - La **mediana** (50%) es 64.1%, lo que sugiere que la distribución tiene sesgo hacia abajo (algunos países con muy bajo acceso reducen la media).
> - El mínimo de porcentaje es **1.7%** (país con casi nulo acceso a internet).
> - El máximo es **100%** (conectividad total).
> - La desviación estándar de `Internet users` es enorme (1.01e+08), lo que refleja la masiva disparidad entre países grandes (China: 1.1 mil millones) y pequeños territorios (mínimo: 1,034 usuarios).

### Paso 30 — Detectar columnas con valores nulos (`isnull().any()`)

In [22]:
df_cib_usenet.isnull().any()

Country or area              False
Subregion                    False
Region                       False
Internet users               False
Population_2021              False
Year                         False
Percentage_Internet_Users    False
dtype: bool

> **Resultado**: Todas las columnas devuelven `False` — **no hay valores nulos** en este dataset. Es un dataset limpio y completo.

### Paso 31 — Contar exactamente los valores nulos por columna (`isnull().sum()`)

In [20]:
df_cib_usenet.isnull().sum()

Country or area              0
Subregion                    0
Region                       0
Internet users               0
Population_2021              0
Year                         0
Percentage_Internet_Users    0
dtype: int64

> **Resultado**: Todos los conteos son **0**. Confirmado: el dataset no tiene ningún valor faltante.

### Paso 32 — Estadísticas de la columna `Percentage_Internet_Users` (`agg()`)

**¿Qué hace esta celda?** Calcula la media, máximo y mínimo del porcentaje de usuarios de internet.

In [23]:
df_cib_usenet['Percentage_Internet_Users'].agg(['mean', 'max', 'min'])

mean     57.368372
max     100.000000
min       1.700000
Name: Percentage_Internet_Users, dtype: float64

> **Interpretación**:
> - El **promedio global** de penetración de internet es **57.37%**.
> - Existe al menos un país con el **100% de penetración** de internet.
> - El país con menor acceso tiene apenas **1.7%** de su población conectada, lo que refleja la enorme brecha digital global.

### Paso 33 — Análisis agrupado por región: usuarios e internet (`groupby()`)

**¿Qué hace esta celda?** Agrupa los países por región y calcula el máximo y mínimo del porcentaje de usuarios de internet y del número total de usuarios de internet.

In [24]:
df_cib_usenet.groupby('Region')[['Percentage_Internet_Users', 'Internet users']].agg(['max', 'min'])

         Percentage_Internet_Users       Internet users        
                               max   min            max     min
Region                                                         
Africa                        73.9   1.7      136203231    2906
Americas                      96.0  11.8      311300000    2833
Asia                          97.8  10.1     1102140000  275717
Europe                       100.0  35.5      129800000   20100
Oceania                       97.6   7.0       25309515    1034

> **Interpretación**:
> - **Europe** tiene el mayor porcentaje máximo de penetración (100%) y el mínimo más alto (35.5%), lo que indica que incluso los países europeos con menor conectividad tienen un nivel aceptable.
> - **Africa** tiene el porcentaje mínimo más bajo (1.7%), confirmando la brecha digital en el continente.
> - **Asia** domina en términos absolutos de usuarios de internet (máximo de 1,102 millones — China), aunque hay una gran disparidad interna.
> - **Oceania** tiene el número mínimo absoluto de usuarios más bajo (1,034), correspondiente a pequeños territorios insulares.

---
## RESUMEN GENERAL DEL EDA

| Dataset | Filas | Columnas | Nulos | Observación clave |
|---|---|---|---|---|
| `df_threads` (PostgreSQL) | Requiere BD activa | — | — | Contiene incidentes de ciberseguridad con pérdidas financieras y tiempos de resolución |
| `df_cib_EDA` (CSV) | 192 | 6 | Sí (CEI: 44% faltante) | Índices globales de ciberseguridad por país; gran brecha entre regiones |
| `df_cib_usenet` (CSV) | 215 | 7 | No | Dataset limpio; brecha digital significativa entre Africa/Asia y Europa |

**Próximos pasos sugeridos:**
1. Tratar los valores nulos del dataset `df_cib_EDA` (imputación o eliminación).
2. Realizar visualizaciones (mapas, gráficos de barras, correlaciones) para explorar relaciones entre los índices de ciberseguridad y el porcentaje de usuarios de internet.
3. Analizar si existe correlación entre el nivel de acceso a internet y el nivel de madurez en ciberseguridad de los países.